File: android/src/main/kotlin/id/passage/passage_flex_flutter/PassageFlexFlutter.kt
package id.passage.passage_flex_flutter

import android.app.Activity
import android.os.Build
import com.google.gson.Gson
import id.passage.android.passageflex.model.AuthenticatorAttachment
import id.passage.passageflex.PassageFlex
import id.passage.passageflex.exceptions.AuthenticateCredentialException
import id.passage.passageflex.exceptions.AuthenticateDiscoverableLoginException
import id.passage.passageflex.exceptions.AuthenticateInvalidRequestException
import id.passage.passageflex.exceptions.RegisterCredentialException
import id.passage.passageflex.exceptions.RegisterInvalidRequestException
import id.passage.passageflex.models.PasskeyCreationOptions
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

internal class PassageFlexFlutter(activity: Activity, appId: String) {

    private val passage: PassageFlex = PassageFlex(activity, appId)

    internal companion object {
        internal fun invalidArgumentError(result: MethodChannel.Result) {
            result.error(
                PassageFlutterError.INVALID_ARGUMENT.name, "Invalid or missing argument", null
            )
        }
    }

    internal fun register(call: MethodCall, result: MethodChannel.Result) {
        val transactionId = call.argument<String>("transactionId") ?: return invalidArgumentError(result)
        val optionsMap = call.argument<Map<String, String>>("passkeyCreationOptions")
        val attachmentValue = optionsMap?.get("authenticatorAttachment")

        val authenticatorAttachment = when (attachmentValue) {
            "platform" -> AuthenticatorAttachment.platform
            "cross-platform" -> AuthenticatorAttachment.crossMinusPlatform
            "any" -> AuthenticatorAttachment.any
            else -> AuthenticatorAttachment.platform
        }
        val passkeyCreationOptions = PasskeyCreationOptions(authenticatorAttachment)

        CoroutineScope(Dispatchers.IO).launch {
            try {
                val nonce = passage.passkey.register(transactionId, passkeyCreationOptions)
                result.success(nonce)
            } catch (e: Exception) {
                val error = when (e) {
                    is RegisterInvalidRequestException -> {
                        PassageFlutterError.INVALID_REQUEST
                    }
                    is RegisterCredentialException -> {
                        PassageFlutterError.WEB_AUTH_FAILED
                    }
                    else -> PassageFlutterError.PASSKEY_ERROR
                }
                result.error(error.name, e.message, e.toString())
            }
        }
    }

    internal fun authenticate(call: MethodCall, result: MethodChannel.Result) {
        val transactionId = call.argument<String>("transactionId") ?: return invalidArgumentError(result)
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val nonce = passage.passkey.authenticate(transactionId)
                result.success(nonce)
            } catch (e: Exception) {
                val error = when (e) {
                    is AuthenticateInvalidRequestException -> {
                        PassageFlutterError.INVALID_REQUEST
                    }
                    is AuthenticateDiscoverableLoginException -> {
                        PassageFlutterError.DISCOVERABLE_LOGIN_FAILED
                    }
                    is AuthenticateCredentialException -> {
                        PassageFlutterError.WEB_AUTH_FAILED
                    }


                    else -> PassageFlutterError.PASSKEY_ERROR
                }
                result.error(error.name, e.message, e.toString())
            }
        }
    }

}

File: ios/Classes/PassageFlexFlutter.swift
import Flutter
import PassageFlex
import AnyCodable

public class PassageFlexFlutter {
    private let passageFlex: PassageFlex

    init(appId: String) {
        passageFlex = PassageFlex(appId: appId)
    }

    func register(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            let error = "UNSUPPORTED_OS_VERSION"
            result(error)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let transactionId = args["transactionId"] as? String else {
            result(PassageFlutterError.invalidArgument.flutterError)
            return
        }
        let optionsMap = args["passkeyCreationOptions"] as? [String: String]
        let attachmentValue = optionsMap?["authenticatorAttachment"]
        let authenticatorAttachment: AuthenticatorAttachment
        switch attachmentValue {
        case "platform":
            authenticatorAttachment = .platform
        case "cross-platform":
            authenticatorAttachment = .crossPlatform
        case "any":
            authenticatorAttachment = .any
        default:
            authenticatorAttachment = .platform
        }

        let passkeyCreationOptions = PasskeyCreationOptions(authenticatorAttachment: authenticatorAttachment)

        Task {
            do {
                let nonce = try await passageFlex.passkey.register(with: transactionId, options: passkeyCreationOptions)
                result(nonce)
            } catch {
                print(error)
                result(mapErrorToFlutter(error))
            }
        }
    }

    func authenticate(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard #available(iOS 16.0, *) else {
            let error = "UNSUPPORTED_OS_VERSION"
            result(error)
            return
        }
        guard let args = call.arguments as? [String: Any],
              let transactionId = args["transactionId"] as? String else {
            result(PassageFlutterError.invalidArgument.flutterError)
            return
        }
        
       Task {
           do {
               let nonce = try await passageFlex.passkey.authenticate(with: transactionId)
               result(nonce)
           } catch {
               result(mapErrorToFlutter(error))
           }
       }
    }

    private func mapErrorToFlutter(_ error: Error) -> FlutterError {
        let passageError: PassageFlutterError
        switch error {
        case let passkeyError as PassagePasskeyAuthorizationError:
            switch passkeyError {
            case .userCanceled:
                passageError = .invalidRequest  
            case .failed:
                passageError = .passkeyError
            case .webauthnError:
                passageError = .webAuthFailed
            case .unknown:
                passageError = .passkeyError
            }
        default:
            passageError = .passkeyError
        }
        return passageError.flutterError
    }

}
